File: chapter_3_stacks_and_queues/js/3.6-animalShelter.js
const { Queue } = require("./queue")

class Animal {
  /* Animal constructor */
  constructor(name) {
    this.name = name
    this._order = 0
  }

  /* Order getter */
  get order() {
    return this._order
  }

  /* Order setter */
  set order(number) {
    this._order = number
  }

  /* Determines if this animal is than another */
  isOlderThan(animal) {
    return this.order < animal.order
  }
}

class Dog extends Animal {
  /* Dog constructor */
  constructor(name) {
    super(name)
  }
}

class Cat extends Animal {
  /* Cat constructor */
  constructor(name) {
    super(name)
  }
}

class AnimalQueue {
  /* Animal queue constructor */
  constructor() {
    this.cats = new Queue()
    this.dogs = new Queue()
    this.order = 0
  }

  /* Add an animal to correct queue */
  enqueue(animal) {
    /* Order is used as a sort of timestamp, so that we can compare
     * the insertion order of a dog to a cat */
    animal.order = this.order
    this.order++
    if (animal instanceof Dog) this.dogs.enqueue(animal)
    else if (animal instanceof Cat) this.cats.enqueue(animal)
  }

  /* Look at tops of dog and cat queues, and pops the queue
   * with the oldest value */
  dequeueAny() {
    if (this.cats.isEmpty()) return this.dequeueDogs()
    if (this.dogs.isEmpty()) return this.dequeueCats()
    return this.dogs.peek().isOlderThan(this.cats.peek())
      ? this.dequeueDogs()
      : this.dequeueCats()
  }

  /* Pops the oldest cat from cat queue */
  dequeueCats() {
    return this.cats.dequeue()
  }

  /* Pops the oldest dog from dog queue */
  dequeueDogs() {
    return this.dogs.dequeue()
  }
}

const aq = new AnimalQueue()
console.log(aq)
/*
AnimalQueue {
  cats: Queue { queue: [], length: 0 },
  dogs: Queue { queue: [], length: 0 },
  order: 0 }
*/
const smokey = new Cat("smokey")
const spot = new Dog("spot")
const garfield = new Cat("garfield")
const grumpy = new Cat("grumpy")
const matador = new Dog("matador")
aq.enqueue(smokey)
aq.enqueue(matador)
aq.enqueue(garfield)
aq.enqueue(grumpy)
aq.enqueue(spot)
console.log(aq)
/*
AnimalQueue {
  cats: Queue { queue: [ [Cat], [Cat], [Cat] ], length: 3 },
  dogs: Queue { queue: [ [Dog], [Dog] ], length: 2 },
  order: 5 }
*/
console.log(aq.dequeueAny()) // Cat { name: 'smokey', _order: 0 }
console.log(aq.dequeueCats()) // Cat { name: 'garfield', _order: 2 }
console.log(aq.dequeueDogs()) // Dog { name: 'matador', _order: 1 }
console.log(aq.dequeueDogs()) // Dog { name: 'spot', _order: 4 }
console.log(aq.dequeueAny()) // Cat { name: 'grumpy', _order: 3 }

File: chapter_3_stacks_and_queues/python/3.6-animal_shelter.py
Queue = __import__('queue').Queue


class Animal:
    def __init__(self, name):
        self.name = name
        self._order = 0

    @property
    def order(self):
        return self._order

    @order.setter
    def order(self, value):
        self._order = value

    def is_older_than(self, animal):
        return self.order < animal.order


class Cat(Animal):
    def __init__(self, name):
        super().__init__(name)


class Dog(Animal):
    def __init__(self, name):
        super().__init__(name)


class AnimalQueue:
    def __init__(self):
        self.dogs = Queue()
        self.cats = Queue()
        self.order = 0

    def enqueue(self, animal):
        """
        Order is used as a sort of timestamp, so that we can compare
        the insertion order of a dog to a cat
        """
        animal.order = self.order
        self.order += 1
        if isinstance(animal, Cat):
            self.cats.enqueue(animal)
        elif isinstance(animal, Dog):
            self.dogs.enqueue(animal)

    def dequeue_any(self):
        """
        Look at tops of dog and cat queues, and pop the queue
        with the oldest value
        """
        if self.cats.is_empty():
            return self.dequeue_dogs()
        if self.dogs.is_empty():
            return self.dequeue_cats
        cat = self.cats.peek()
        dog = self.dogs.peek()
        if cat.is_older_than(dog):
            return self.dequeue_cats()
        else:
            return self.dequeue_dogs()

    def dequeue_cats(self):
        return self.cats.dequeue()

    def dequeue_dogs(self):
        return self.dogs.dequeue()


if __name__ == "main":
    aq = AnimalQueue()
    spot = Dog('spot')
    smokey = Cat('smokey')
    grumpy = Cat('grumpy')
    matador = Dog('matador')
    garfield = Cat('garfield')
    aq.enqueue(spot)
    aq.enqueue(smokey)
    aq.enqueue(grumpy)
    aq.enqueue(matador)
    aq.enqueue(garfield)
    print([x.name for x in aq.cats.queue], [x.name for x in aq.dogs.queue])
    # ['smokey', 'grumpy', 'garfield'] ['spot', 'matador']
    print(aq.dequeue_any().name)  # spot
    print(aq.dequeue_cats().name)  # smokey
    print(aq.dequeue_any().name)  # grumpy
